client/tinychain/ml/optimizer.py

```diff
@@ -13,29 +14,54 @@ class Optimizer(Map):
     def optimize(self, i, gradient, inputs, loss):
         """Update the given `gradient` by computing deltas for the given `loss`."""
 
+    @property
+    def lr(self) -> F32:
+        return self['lr']
+
 
 class GradientDescent(Optimizer):
     @classmethod
-    def create(cls, learning_rate=F32(0.01)):
+    def create(cls, lr=F32(0.01)):
         """Create a new `GradientDescent` optimizer with the given `learning_rate`."""
 
-        return cls({"lr": learning_rate})
+        return cls({"lr": lr})
 
-    def optimize(self, i, gradient, inputs, loss):
-        lr = self["lr"]
+    def optimize(self, i, param_list: List[Parameter]):
+        return [param.value.write(param.value - (param.grad * self.lr)) for param in param_list]
+
+
+class Adam(Optimizer):
+    @classmethod
+    def create(cls, beta1=F32(0.9), beta2=F32(0.999), lr=F32(1e-3), eps=F32(1e-8)):
+        return cls({'beta1': beta1, 'beta2': beta1, 'lr': lr, 'eps': eps})
+
+    @property
+    def beta1(self) -> F32:
+        return self['beta1']
```

Comment:
You should be able to access `self['beta1']` from within the instance methods of the `Adam` class with the type information preserved (as of commit 83ec5c0), so I don't think you need these `@property`s